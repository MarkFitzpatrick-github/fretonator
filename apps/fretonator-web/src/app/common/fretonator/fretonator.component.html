<ng-template #string let-stringName="stringName" let-string="string">
  <ng-container *ngFor="let fret of frets">
    <div
      class="fretboard__cell"
      [class.fretboard__cell--string]="fret === 0"
      [class.fretboard__cell--selected]="fretMap | getFretFromFretMap: string:fret"
      [attr.data-string-name]="fret === 0 ? stringName : null"
      [attr.data-string]="string"
      [attr.data-fret]="fret"
      [attr.data-degree]="(fretMap | getFretFromFretMap: string : fret)?.degree"
      [attr.data-display-note]="(fretMap | getFretFromFretMap: string : fret)?.displayName"
      [attr.data-mode]="mode"
      (click)="playbackService.playNote(stringName, fret)"
    ></div>
  </ng-container>
</ng-template>

<div class="fretboard__toggle">
  <button class="fretboard__toggleButton fretboard__toggleButton--left"
          [class.fretboard__toggleButton--active]="fretMode === fretModes.twelve"
          (click)="setFretMode(fretModes.twelve)">12 frets
  </button>
  <button class="fretboard__toggleButton fretboard__toggleButton--right"
          [class.fretboard__toggleButton--active]="fretMode === fretModes.twentyFour"
          (click)="setFretMode(fretModes.twentyFour)">24 frets
  </button>
</div>

<div class="arrowHint">
  <span class="arrowHint__arrow"></span>
</div>

<div class="fretboard__container"
     [class.fretboard__wide]="fretMode === fretModes.twentyFour">

  <div class="fretboard"
       [class.fretboard__24]="fretMode === fretModes.twentyFour">
    <ng-container
      [ngTemplateOutlet]="string"
      [ngTemplateOutletContext]="{ stringName: 'e', string: 'E' }"
    ></ng-container>

    <ng-container
      [ngTemplateOutlet]="string"
      [ngTemplateOutletContext]="{ stringName: 'B', string: 'B' }"
    ></ng-container>

    <ng-container
      [ngTemplateOutlet]="string"
      [ngTemplateOutletContext]="{ stringName: 'G', string: 'G' }"
    ></ng-container>

    <ng-container
      [ngTemplateOutlet]="string"
      [ngTemplateOutletContext]="{ stringName: 'D', string: 'D' }"
    ></ng-container>

    <ng-container
      [ngTemplateOutlet]="string"
      [ngTemplateOutletContext]="{ stringName: 'A', string: 'A' }"
    ></ng-container>

    <ng-container
      [ngTemplateOutlet]="string"
      [ngTemplateOutletContext]="{ stringName: 'E', string: 'E' }"
    ></ng-container>
  </div>
</div>

<h4 class="fretonator__playCta">
  <span class="fretonator__playCtaIcon">
    <app-speaker-svg></app-speaker-svg>
  </span>
  Tap notes on the fretboard for sound!
</h4>

<div class="infoContainer">
  <app-scale-map [mode]="mode"
                 [modeMap]="modeMap"
                 [isTheoretical]="modeMap | isTheoreticalScale"
                 [modeDisplayString]="modeDisplayString"
                 [note]="note"
                 [noteExtenderString]="noteExtenderString"></app-scale-map>

  <app-interval-map [intervalMap]="intervalMap"></app-interval-map>

  <app-chord-map [chordMap]="chordMap"></app-chord-map>
</div>
