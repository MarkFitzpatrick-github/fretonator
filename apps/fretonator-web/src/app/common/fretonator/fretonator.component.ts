import { Component, Input } from '@angular/core';
import { ChordMap, FretMap, Mode, ModeMap, Scale } from '../../util/types';
import { NotePlaybackService } from '../playback/note-playback.service';

@Component({
  selector: 'app-fretonator',
  templateUrl: './fretonator.component.html',
  styleUrls: ['./fretonator.component.scss'],
})
export class FretonatorComponent {
  @Input() fretMap: FretMap;
  @Input() modeMap: ModeMap;
  @Input() scale: Scale;
  @Input() intervalMap: [];
  @Input() chordMap: ChordMap;
  @Input() mode: Mode;
  @Input() modeDisplayString: string;
  @Input() note: string;
  @Input() noteExtenderString: string;

  constructor(public playbackService: NotePlaybackService) { }

  frets = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];
}
