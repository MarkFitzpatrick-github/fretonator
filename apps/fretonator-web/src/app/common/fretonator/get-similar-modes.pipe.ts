import { Pipe, PipeTransform } from '@angular/core';
import { FretMapService } from '../fret-map/fret-map.service';
import { Mode, ModeMap, SimilarModes } from '../../util/types';

@Pipe({
  name: 'getSimilarModes'
})
export class GetSimilarModesPipe implements PipeTransform {

  constructor(public generateFretMapService: FretMapService) {}

  transform(modeMap: ModeMap, mode: Mode): SimilarModes {
    return this.generateFretMapService.getSimilarModes(modeMap, mode);
  }

}
