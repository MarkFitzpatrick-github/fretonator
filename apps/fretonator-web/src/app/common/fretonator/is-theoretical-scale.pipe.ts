import { Pipe, PipeTransform } from '@angular/core';
import { ModeMap } from '../../util/types';

@Pipe({
  name: 'isTheoreticalScale'
})
export class IsTheoreticalScalePipe implements PipeTransform {

  transform(modeMap: ModeMap): boolean {
    for (const note of modeMap) {
      if(note.displayName.indexOf('x') !== -1 || note.displayName.indexOf('ùÑ´') !== -1) {
        return true;
      }
    }
    return false;
  }
}
