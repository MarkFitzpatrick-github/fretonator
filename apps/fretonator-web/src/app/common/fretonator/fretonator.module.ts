import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FretonatorComponent } from './fretonator.component';
import { GetFretFromFretMapPipe } from './get-fret-from-fret-map.pipe';
import { ChordMapModule } from './chord-map/chord-map.module';
import { IntervalMapModule } from './interval-map/interval-map.module';
import { ScaleMapModule } from './scale-map/scale-map.module';
import { IsTheoreticalScalePipe } from './is-theoretical-scale.pipe';
import { SpeakerModule } from '../svgs/speaker/speaker.module';
import { SimilarModesModule } from './similar-modes/similar-modes.module';
import { GetSimilarModesPipe } from './get-similar-modes.pipe';
import { VideoLoaderModule } from '../video-loader/video-loader.module';
import { SymbolToNoteObjPipe } from './symbol-to-note-obj.pipe';
import { GetJamTrackPipe } from './get-jam-track.pipe';

@NgModule({
  declarations: [
    FretonatorComponent,
    GetFretFromFretMapPipe,
    IsTheoreticalScalePipe,
    GetSimilarModesPipe,
    SymbolToNoteObjPipe,
    GetJamTrackPipe
  ],
  imports: [
    CommonModule,
    ChordMapModule,
    IntervalMapModule,
    ScaleMapModule,
    SpeakerModule,
    SimilarModesModule,
    VideoLoaderModule
  ],
  exports: [
    FretonatorComponent
  ]
})
export class FretonatorModule {
}
