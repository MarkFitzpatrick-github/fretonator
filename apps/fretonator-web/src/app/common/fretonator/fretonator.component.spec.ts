import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { FretonatorComponent } from './fretonator.component';
import { FretonatorModule } from './fretonator.module';
import { Component, DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';

describe('FretonatorComponent', () => {

  const selectors = {
    twelveButton: By.css('.fretboard__toggleButton--twelve'),
    twentyFourButton: By.css('.fretboard__toggleButton--twentyFour')
  };

  const classNames = {
    toggleFretButtonSelected: 'fretboard__toggleButton--active'
  };

  @Component({
    selector: 'app-fretonator-spec',
    template: `
      <app-fretonator [modeMap]="[{
                                  name: 'f',
                                  sharp: false,
                                  flat: false,
                                  doubleFlat: false,
                                  doubleSharp: false,
                                  displayName: 'F'
                                }]"
                      [fretMap]="[
                      { string: 'A', fret: 0, displayName: 'A', degree: 'dominant' },
                      { string: 'A', fret: 2, displayName: 'B', degree: 'submediant' }]">
      </app-fretonator>
    `
  })
  class FretonatorComponentSpec {
  }

  let component: FretonatorComponentSpec;
  let fixture: ComponentFixture<FretonatorComponentSpec>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [FretonatorComponentSpec],
      imports: [FretonatorModule]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(FretonatorComponentSpec);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('setFretMode()', () => {
    let twelveButton: DebugElement;
    let twentyFourButton: DebugElement;

    beforeEach(() => {
      twelveButton = fixture.debugElement.query(selectors.twelveButton);
      twentyFourButton = fixture.debugElement.query(selectors.twentyFourButton);
    });

    it('changes the fret mode to twenty four when the button is clicked', () => {
      twentyFourButton.nativeElement.click();
      fixture.detectChanges();
      expect(twentyFourButton.classes[classNames.toggleFretButtonSelected]).toBeTruthy();
    });

    it('changes the fret mode to twelve when the button is clicked', () => {
      twelveButton.nativeElement.click();
      fixture.detectChanges();
      expect(twelveButton.classes[classNames.toggleFretButtonSelected]).toBeTruthy();
    });
  });
});
