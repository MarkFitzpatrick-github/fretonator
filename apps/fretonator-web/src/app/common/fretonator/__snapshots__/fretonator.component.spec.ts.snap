// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FretonatorComponent should create 1`] = `FretonatorComponentSpec {}`;
